### PR TITLE
Filter the undefined and empty values before making the request to filter nodes and farms.

### PR DESCRIPTION
**Pull request by Mahmoud-Emad**
### Description

Filter the empty/undefined values from the object

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/2353

### Checklist

- [ ] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

**Comment by amiraabouhadid** [comment]
farm requests still have undefined fields 
image

File: packages/grid_client/src/primitives/nodes.ts [hunk 1]
```diff
@@ -404,12 +404,19 @@ class Nodes {
       region: options.region,
       healthy: options.healthy,
     };
+
```
Comment (amiraabouhadid) [inline packages/grid_client/src/primitives/nodes.ts:407]: please extract this, place it into utils and apply it to farms too

**Review by amiraabouhadid — CHANGES_REQUESTED** [review]

**Review by amiraabouhadid — CHANGES_REQUESTED** [review]
please avoid repeated code. apply filter to farms too

File: packages/grid_client/src/primitives/nodes.ts [hunk 2]
```diff
@@ -438,7 +441,9 @@ class Nodes {
       ret_count: options.ret_count,
       region: options.region,
     };
-    return Object.entries(params)
+
+    const filteredParams = removeUndefinedAndEmptyString(params);
+    return Object.entries(filteredParams)
```
Comment (amiraabouhadid) [inline packages/grid_client/src/primitives/nodes.ts]: why not extract Object.entries() etc also into the function in the utils? Seems unnecessarily repetitive still

**Review by amiraabouhadid — APPROVED** [review]

File: yarn.lock [hunk 3]
```diff
@@ -20225,7 +20192,7 @@ typescript@*, typescript@^5.0.2:
   resolved "https://registry.yarnpkg.com/typescript/-/typescript-5.0.4.tgz#b217fd20119bd61a94d4011274e0ab369058da3b"
   integrity sha512-cW9T5W9xY37cc+jfEnaUvX91foxtHkza3Nw3wkoF4sSlKn0MONdkdEndig/qPBWXNkmplh3NzayQzCiHM4/hqw==
 
-"typescript@^3 || ^4", typescript@^4.0.0, typescript@^4.4.4, typescript@^4.7.4, typescript@^4.8.4, typescript@^4.9.3:
+"typescript@^3 || ^4", typescript@^4.4.4, typescript@^4.7.4, typescript@^4.8.4, typescript@^4.9.3:
```
Comment (zaelgohary) [inline yarn.lock:20195]: Why are there yarn changes?

File: packages/grid_client/src/helpers/utils.ts [hunk 4]
```diff
@@ -63,6 +63,25 @@ function zeroPadding(length: number, num: number) {
...
+ */
+function convertObjectToQueryString(obj: Record<string, any>): string {
+  // Filter out undefined or empty string values
+  const filteredParams = Object.fromEntries(
+    Object.entries(obj).filter(([_, value]) => value !== undefined && value !== ""),
+  );
+
+  // Convert object to query string format
+  const queryString = Object.entries(filteredParams)
+    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
```
Comment (zaelgohary) [inline packages/grid_client/src/helpers/utils.ts:79]: Why do we use encodeURIComponent?

Comment (Mahmoud-Emad) [reply packages/grid_client/src/helpers/utils.ts:79]: Its a built in helper used to encode special characters in a URL.

e.g.

```ts
const searchTerm = "hello world!";
const url = "https://example.com/search?q=" + encodeURIComponent(searchTerm);
console.log(url);
// Output: https://example.com/search?q=hello%20world%21
```

**Review by zaelgohary — APPROVED** [review]
Good job ya Omda.

File: packages/grid_client/src/helpers/utils.ts [hunk 5]
```diff
@@ -63,6 +63,25 @@ function zeroPadding(length: number, num: number) {
   return num.toString(16).padStart(length, "0");
 }
 
+/**
+ * Converts an object into a query string format.
+ * @param obj - The object to be converted into a query string.
+ * @returns A string representing the object in query string format.
+ */
+function convertObjectToQueryString(obj: Record<string, any>): string {
```
Comment (AhmedHanafy725) [inline packages/grid_client/src/helpers/utils.ts:71]: please add unit tests for this method

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

File: packages/grid_client/src/helpers/utils.ts (hunk 5 above)
Comment (Mahmoud-Emad) [reply packages/grid_client/src/helpers/utils.ts:71]: The function has been added.

File: packages/grid_client/tests/functions/query_ot_string.test.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import { convertObjectToQueryString } from "../../src/helpers/utils";
+
+describe("convertObjectToQueryString", () => {
```
Comment (AhmedHanafy725) [inline packages/grid_client/tests/functions/query_ot_string.test.ts:3]: why do we use `camelcase` for the test description?

File: packages/grid_client/tests/test-reports/report.xml
Comment (AhmedHanafy725) [inline packages/grid_client/tests/test-reports/report.xml:1]: please remove the test result report from the PR

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]
let's rename the `functions` directory to `unittests`, and filename from `query_ot_string.test.ts` to `query_to_string.test.ts`

File: packages/grid_client/tests/functions/query_ot_string.test.ts (hunk 6 above)
Comment (Mahmoud-Emad) [reply packages/grid_client/tests/functions/query_ot_string.test.ts:3]: Updated

**Review by AhmedHanafy725 — APPROVED** [review]

**Pull request merged**